Book/php7/zend_engine/zend_opcache.rst

```diff
@@ -1,2 +1,104 @@
+following exceptions:
+
+* ``zend_persistent_script.dynamic_members`` stores information about the state of the cached script
+* ``zend_class_entry.inheritance_cache`` is extended if a new combination of parent class and interfaces is encountered
+* ``zend_op.handler`` can be replaced at runtime by the tracing JIT to start executing JITted code
+* ``zend_string.gc.refcount`` for persistent strings can be replaced with map ptr offset to cache class lookups by name
+
+map_ptr
+-------
+
+As mentioned in the introduction, SHM is shared for all child processes. Thus, no changes can be made to the data
+structures living in SHM unless 1. the changes are supposed to be reflected in all child processes and 2. the SHM
+segment has been locked before any changes are written to avoid data races.
+
+Sometimes the data structures contain a field that should be different per process or request. The ``map_ptr`` mechanism
+can be used to achieve this. In short, instead of storing a pointer to per-process data directly (as that would affect
+all processes but each having different addresses for this process specific data), a unique offset is assigned during
+compilation. During runtime, local memory is allocated to hold enough space for all the entries referenced in these
+offsets. Each offset then corresponds to an item in the local memory segment without needing to know the per-process
+location of the exact element.
+
+The map ptr can also store real pointers. This can be useful when the data structure doesn't live in SHM and thus
+doesn't need to store the value in a separate map. The fact that pointers are aligned by their size is used to
+differentiate between offsets and real pointers. The offsets start at 1, 9, 17, etc (assuming a 64-bit pointer size).
+However, pointers (unless padding was removed) would not get aligned this way. Instead, they would be stored in
+addresses that end with 0x0 or 0x8 given that they are 8-byte aligned. The macro ``ZEND_MAP_PTR_IS_OFFSET`` will be used
+internally to check if the ``0b1`` byte is set. If it is, the value stored is an offset. Otherwise it's a direct
+pointer.
+
+Here is an example from php-src.
+
+.. image:: ./images/map_ptr.png
+   :align: center
+
+Classes can contain data that needs to be evaluated at runtime, called ``mutable_data``. This ``mutable_data`` contains
+fields like ``default_properties_table``, ``constants_table`` and more.  ``ZEND_MAP_PTR_DEF`` can be used to declare a
+field that holds the offset into local memory or direct pointer.
+
+::
+
+    struct _zend_class_entry {
+        // ...
+        ZEND_MAP_PTR_DEF(zend_class_mutable_data*, mutable_data);
+        // ...
+    };
+
+``ZEND_MAP_PTR_INIT`` assigns a value directly to the underlying offset field. This must only be done when the structure
+is not yet in SHM. In practice, this marco should usually only be used to initialize the map ptr to ``NULL``.
```

Comment:
marco -> macro. I also wouldn't say this writes the offset field, it writes the pointer. It does not need to be NULL for the non-SHM use case.

Reply:
I tried to point to the fact that usually code should be agnostic on whether an offset or pointer is stored, so you should usually do `ZEND_MAP_PTR_INIT(x->y, NULL)` and then `ZEND_MAP_PTR_SET(x->y, ptr)`. Maybe I can clarify that `ZEND_MAP_PTR_INIT` can be used when you know the structure doesn't live in SHM as a performance optimization.